### ISSUE TITLE
Installing Conan packages (to be build from sources and pre-built) from conanfile.txt

### ISSUE THREAD

**Issue by himansu2906** [conan-io/conan#12254]
We have package dependencies(internal libraries and pre-build external libraries) in the project.

How can we mention both kind of libraries in one conanfile.txt so that pre-built externals (where we don't have sources) are only getting installed and internal can be build from the sources if needed?

means how can we differentiate to build from sources or simply download in conanfile.txt(which is containing all the library dependencies )?

**Comment by memsharded** [comment]
Hi @himansu2906 

Packages created from pre-compiled binaries with ``conan export-pkg`` can specify ``build_policy = "never"``, so a ``--build=missing`` will not affect them.

**Comment by himansu2906** [comment]
@memsharded : thank you very much for the quick answer.

I have one more question that can we create conan package setting independent(os, arch , compiler) for prebuilt libraries?

**Comment by memsharded** [comment]
> I have one more question that can we create conan package setting independent(os, arch , compiler) for prebuilt libraries?

Yes, absolutely. ``conan export-pkg`` works with profiles, settings options, and the recipe itself should declare ``settings = "os", "arch", ....`` as necessary. The ``package()`` method can use conditional over ``if self.settings.os == ....`` to decide to package some prebuilt libraries or others (typically you will have them in different folders). Conan will store as many independent package binaries as necessary for the configurations you manage.

**Comment by himansu2906** [comment]
@memsharded : thank you very much for the answer.

we have one scenario where we are mentioning external and internal libraries in conanfile.txt, please find content below.

[requires]
module_concept/1.3@user/testing  (internal)
libeosSRC36_20_31/1.0 (external)
[generators]
cmake

while installing we are trying to install with different profile that with which these libraries were build, please find below command which we are using to install:

conan install  <path of conanfile.txt> --build=missing -pr=test
note : test profile is having different settings.

Issue: while installing it is showing masses that packages are missing for external since these were build with **build_policy=never** and for internal it is showing build but it is not building the internal libraries from sources.

![image](https://user-images.githubusercontent.com/109611178/195317926-c9c5509f-5c31-4d1d-9c79-0576728f9fb5.png)

**observation:** if conanfile.txt is having only internal library, then it is building from sources. 

conanfile.txt :

[requires]
module_concept/1.3@user/testing
[generators]
cmake

output while calling same command:

![image](https://user-images.githubusercontent.com/109611178/195318846-31f0ae03-d0bc-4ed2-b3af-223d8a913d49.png)

**how can we resolve this issue since we want to put all the dependencies (external and interna) together l in conanfile.txt?**

**Comment by memsharded** [comment]
If you are using that package from pre-compiled binaries, you need to create the different binaries for the different profiles too, like:

```bash
conan export-pkg . -pr=myprofile
conan export-pkg . -pr=test
...
```
Typically, each one will result in different binaries being packaged, with the ``package()`` method having conditionals on settings to pick some or others artifacts from the precompiled binaries.

**Comment by himansu2906** [comment]
@memsharded : our primary motive is to create internal libraries from sources if binaries are missing and external can be skipped from the list(conanfile.txt).

we want to build packages(if binaries are missing) from sources while installing it.

**Comment by memsharded** [comment]
Sorry, I don't get what you mean. If the packages are being created with ``conan export-pkg`` from pre-compiled binaries, Conan doesn't know how to build those packages from sources. You need to pre-emptively create all the different required binaries, running ``conan export-pkg`` for the different configurations/profiles that you will need. Only packages created with ``conan create`` that contain the ``build()`` method specifying how to build from source, can be built with ``--build=missing``.

**Comment by himansu2906** [comment]
we have 2 kind of libraries.
external : which are pre-compiled banaries and provided by external companies. we don't have source for these libraries
internal : which are being created from sources by us.

let say lib1(external) and lib2(internal) are already created with profile1 and uploaded to artifactory.

Both libs are mentioned in conanfile.txt as a dependency for a project.

now when we are consuming these libraries with different profile2 and "--build=missing", expectation is that externals can be skipped since different profile but internal can be created locally from the sources with profile2.

**Comment by memsharded** [comment]
But it doesn't work that way, it doesn't matter if libraries are internal or internal. You cannot just apply a different profile that says ``os=Windows`` and ``arch=x86`` and expect the pre-compiled binary package created with ``conan export-pkg`` to work. 

So 2 things can happen:

- You assume that the precompiled binary is good for a range of different settings, and it will be binary compatible. In that case, you can remove the ``compiler`` and ``build_type``, for example, from the settings leaving ``settings = "os", "arch"`` in the recipe.
- The provided precompiled libraries are different for different settings, different compiler versions or build type, then you need to pre-emptively create all the different configurations with ``conan export-pkg ... --profile=`` with all the different profiles (profile1, profile2) that will be used later

**Comment by himansu2906** [comment]
it works perfectly fine when we consume the package individually with different profile, it builds the binary from sources locally.

> conan install  module_concept/1.3@user/testing  --build=missing -pr=test

![image](https://user-images.githubusercontent.com/109611178/195340646-d627fbc8-cf9b-40d7-a089-da894a237274.png)

**Comment by memsharded** [comment]
Yes, the ``module_concept`` is internal, it knows how to build from sources, typically contain a ``build()`` method, and everything works.
If I understand correctly, the problem is with the other package, the external ``libeosSRC36_20_31``, which doesn't contain a ``build()`` method because it is created from pre-compiled binaries with ``conan export-pkg``. This package cannot be built from source with ``--build missing`` because it doesn't know how to build from sources. This is the package that you either:
- Control the declared settings so they are valid for different configurations
- Create the different necessary binaries for the different profiles with multiple ``conan export-pkg`` commands

**Comment by himansu2906** [comment]
Thank you very much @memsharded , it clarifies or doubt.

**Comment by memsharded** [comment]
Closing this old issue as solved, please create new tickets for any further question, thanks!
